### PR TITLE
Permute receivers (uniformly) randomly

### PR DESCRIPTION
**Pull request by akalin-keybase**
This prevents recipients from being able
to infer data about other recipients based
on order.

### CODE REVIEW

File: signcrypt_seal_test.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+package saltpack
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestSealEncryptionKeyForReceiversPermuted(t *testing.T) {
+	const count = 10
```
Comment (akalin-keybase) [inline signcrypt_seal_test.go]: TODO: Flesh this test out a bit more

File: signcrypt_seal.go [hunk 2]
```diff
@@ -246,6 +236,24 @@ func (sss *signcryptSealStream) init(receiverBoxKeys []BoxPublicKey, receiverSym
...
 
+func sealEncryptionKeyForReceivers(receiverBoxKeys []BoxPublicKey, receiverSymmetricKeys []ReceiverSymmetricKey, ephemeralKey BoxSecretKey, encryptionKey SymmetricKey) []receiverKeys {
+	order := randomPerm(len(receiverBoxKeys) + len(receiverSymmetricKeys))
+
+	receiverKeysArray := make([]receiverKeys, len(receiverBoxKeys)+len(receiverSymmetricKeys))
+	// Collect all the recipient identifiers, and encrypt the payload key for
+	// all of them.
+	for i, receiverBoxKey := range receiverBoxKeys {
+		index := order[i]
+		receiverKeysArray[index] = receiverEntryForBoxKey(receiverBoxKey, ephemeralKey, encryptionKey, uint64(index))
```
Comment (akalin-keybase) [inline signcrypt_seal.go]: Ah, the identifier depends on the index, unlike the encryption case. Does this imply quadratic behavior on the decryption side, like you mentioned?

**Comment by akalin-keybase** [comment]
Looks like the coverage decrease is due to some added panics for switch cases.

File: encrypt.go [hunk 3]
```diff
@@ -187,17 +175,64 @@ func (es *encryptStream) init(version Version, sender BoxSecretKey, receivers []
...
+func computeReceiverOrder(version Version, receiverCount int) []int {
+	switch version.Major {
+	case 1:
+		order := make([]int, receiverCount)
+		for i := 0; i < receiverCount; i++ {
+			order[i] = i
+		}
+		return order
+	case 2:
+		return randomPerm(receiverCount)
```
Comment (akalin-keybase) [inline encrypt.go]: Oh duh does it even make sense to permute the order for encryption mode? The IDs are already public, right?

Comment (oconnor663) [inline encrypt.go]: By default yes, though there's an anonymous mode that could benefit.

File: signcrypt_seal.go (hunk 2 above)
Comment (oconnor663) [inline signcrypt_seal.go]: I think so, though currently the number of receiver box secret keys you're going to have is one.

Comment (oconnor663) [inline signcrypt_seal.go]: And even when it's more than one (shared DH keys?), the number of recipients on a message is also expected to be "small".

File: common.go [hunk 4]
```diff
@@ -6,10 +6,11 @@ package saltpack
 import (
 	"bytes"
 	"crypto/hmac"
-	"crypto/rand"
+	cryptorand "crypto/rand"
```
Comment (oconnor663) [inline common.go]: This is a good idea. I'll start doing it elsewhere.

File: common.go [hunk 5]
```diff
@@ -37,6 +38,27 @@ func randomFill(b []byte) (err error) {
...
+
+var _ mathrand.Source = cryptoSource{}
+
+// No need to implement Source64, since mathrand.Rand.Perm() doesn't use it.
+
+func (s cryptoSource) Int63() int64 {
+	var buf [8]byte
+	cryptorand.Read(buf[:])
+	return int64(binary.BigEndian.Uint64(buf[:]) >> 1)
+}
```
Comment (oconnor663) [inline common.go]: Curious what happens if this function actually returns an int > 2^63...

File: signcrypt_seal_test.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	}
+
+	ephemeralKey := boxSecretKey{key: RawBoxKey{0x08}}
+	encryptionKey := SymmetricKey{0x6}
+
+	receiverKeysArray1 := sealEncryptionKeyForReceivers(receiverBoxKeys, receiverSymmetricKeys, ephemeralKey, encryptionKey)
+	receiverKeysArray2 := sealEncryptionKeyForReceivers(receiverBoxKeys, receiverSymmetricKeys, ephemeralKey, encryptionKey)
+
+	// Technically this check is flaky, but the flake probability
+	// is 1/20! ~ 2^{-61}.
```
Comment (oconnor663) [inline signcrypt_seal_test.go]: lol

File: common.go (hunk 4 above)
Comment (akalin-keybase) [inline common.go]: I did it here out of necessity (since I use both) but yeah maybe it's a good idea in general.

File: common.go (hunk 5 above)
Comment (akalin-keybase) [inline common.go]: You mean a negative number? I think callers in `math.rand` would panic.

File: encrypt.go [hunk 7]
```diff
@@ -189,15 +176,57 @@ func (es *encryptStream) init(version Version, sender BoxSecretKey, receivers []
...
+	// TODO: Plumb the pre-computed shared ephemeral keys from
+	// boxPayloadKeyForReceivers through to computeMACKeysSender.
+	es.macKeys = computeMACKeysSender(es.header.Version, order, sender, ephemeralKey, receivers, es.headerHash)
 
 	return nil
 }
 
+func computeReceiverOrder(version Version, receiverCount int) []int {
+	switch version.Major {
+	case 1:
```
Comment (oconnor663) [inline encrypt.go]: I don't think randomizing the receivers will break anything in v1. Any reason not to do it for both?

File: signcrypt_seal.go [hunk 8]
```diff
@@ -250,6 +240,24 @@ func (sss *signcryptSealStream) init(receiverBoxKeys []BoxPublicKey, receiverSym
...
+	}
+	for i, receiverSymmetricKey := range receiverSymmetricKeys {
+		index := order[len(receiverBoxKeys)+i]
+		receiverKeysArray[index] = receiverEntryForSymmetricKey(receiverSymmetricKey, ephemeralKey.GetPublicKey(), encryptionKey, uint64(index))
+	}
```
Comment (oconnor663) [inline signcrypt_seal.go]: Would it be easier to try to define a function that permuted an array in place? Even if we had to copy it for different data types?

Comment (akalin-keybase) [inline signcrypt_seal.go]: Yeah, that turns out to be much simpler! I implemented that.

File: encrypt.go (hunk 7 above)
Comment (akalin-keybase) [inline encrypt.go]: Yeah, you're right. Done.

**Comment by akalin-keybase** [comment]
Addressed comments, which pretty much caused a rewrite, although it's much simpler now!

I think it's worth you taking another quick review. Thanks!

File: signcrypt_seal.go [hunk 9]
```diff
@@ -180,16 +190,29 @@ func receiverEntryForSymmetricKey(receiverSymmetricKey ReceiverSymmetricKey, eph
...
 	return receiverKeys{
-		ReceiverKID:   receiverSymmetricKey.Identifier,
+		ReceiverKID:   r.Identifier,
 		PayloadKeyBox: payloadKeyBox,
 	}
 }
 
+func shuffleSigncryptionReceivers(receiverBoxKeys []BoxPublicKey, receiverSymmetricKeys []ReceiverSymmetricKey) []receiverKeysMaker {
+	order := randomPerm(len(receiverBoxKeys) + len(receiverSymmetricKeys))
+	receivers := make([]receiverKeysMaker, len(receiverBoxKeys)+len(receiverSymmetricKeys))
```
Comment (oconnor663) [inline signcrypt_seal.go]: could compute `totalLen` once maybe, to make it easier to read?

**Pull request merged**